### PR TITLE
tree: add missing div-by-zero error for float div

### PR DESCRIPTION
**Pull request by jordanlewis**
Previously, dividing a float by 0 would return -Inf, instead of
returning a divide by 0 error as expected.

Release note (bug fix): dividing floats by zero now returns an error,
instead of -Inf.

### CODE REVIEW

**Review by rafiss — COMMENTED** [review]
> Reviewed 2 of 2 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/37774)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis and @mjibson)

___
*[pkg/sql/sem/tree/eval_test.go, line 255 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/37774#-Lf_PcWo6jrpiLej5G3k:-Lf_PcWo6jrpiLej5G3l:b-b23duq) ([raw file](https://github.com/cockroachdb/cockroach/blob/060e03d038dff357e96283889bb7836fe7f78717/pkg/sql/sem/tree/eval_test.go#L255)):*
> ```Go
> 		{`1 % 0`, `zero modulus`},
> 		{`1 / 0`, `division by zero`},
> 		{`1::float / 0::float`, `division by zero`},
> ```

why does the test go here instead of in `testdata/eval/float_decimal_div`? is it because it's an error?

**Review by jordanlewis — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/37774)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mjibson and @rafiss)

___
*[pkg/sql/sem/tree/eval_test.go, line 255 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/37774#-Lf_PcWo6jrpiLej5G3k:-Lf_QBjOBhaTMfQss2ov:b1ox0fw) ([raw file](https://github.com/cockroachdb/cockroach/blob/060e03d038dff357e96283889bb7836fe7f78717/pkg/sql/sem/tree/eval_test.go#L255)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

why does the test go here instead of in `testdata/eval/float_decimal_div`? is it because it's an error?
</blockquote></details>

Yes, as far as I can tell, the `datadriven` package that we use for the eval tests here doesn't know how to expect errors yet. That would be a nice feature addition - for now, this other test has a good way of specifying expected errors.

**Review by jordanlewis — COMMENTED** [review]
Also had to remove 2 older tests that expected float div-by-0 to return non-errors.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/37774#-:-Lf_R_tiA4dkLFjVwpc4:b-q7ujt2)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mjibson and @rafiss)

**Review by rafiss — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/37774)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mjibson)

File: pkg/sql/sem/tree/eval.go [hunk 1]
```diff
@@ -1116,7 +1116,11 @@ var BinOps = map[BinaryOperator]binOpOverload{
 			RightType:  types.Float,
 			ReturnType: types.Float,
 			Fn: func(_ *EvalContext, left Datum, right Datum) (Datum, error) {
-				return NewDFloat(*left.(*DFloat) / *right.(*DFloat)), nil
+				r := *right.(*DFloat)
+				if float64(r) == 0.0 {
```
Comment (mjibson) [inline pkg/sql/sem/tree/eval.go]: I believe this cast is unnecessary?

**Review by madelynnblue — APPROVED** [review]

File: pkg/sql/sem/tree/eval.go (hunk 1 above)
Comment (jordanlewis) [reply pkg/sql/sem/tree/eval.go]: Right you are, done.

**Comment by jordanlewis** [comment]
bors r+

TFTRs!

**Pull request merged**